ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/platform/server/api/services/grpc/GrpcHealthServiceSpec.scala

```diff
@@ -31,24 +29,253 @@ final class GrpcHealthServiceSpec
+        new HealthChecks(
+          "component A" -> healthyComponent,
+          "component B" -> healthyComponent,
+          "component C" -> healthyComponent,
+        ))
+
+      val response = Await.result(service.check(allServicesRequest), patienceConfig.timeout)
 
       response should be(servingResponse)
     }
 
+    "report NOT_SERVING if a single check is unhealthy" in {
+      val service = new GrpcHealthService(
+        new HealthChecks(
+          "component A" -> healthyComponent,
+          "component B" -> unhealthyComponent,
+          "component C" -> healthyComponent,
+        ))
+
+      val response = Await.result(service.check(allServicesRequest), patienceConfig.timeout)
+
+      response should be(notServingResponse)
+    }
+
+    "report SERVING when querying a single, healthy component" in {
+      val service = new GrpcHealthService(new HealthChecks("component" -> healthyComponent))
+
+      val response =
+        Await.result(service.check(serviceRequestFor("component")), patienceConfig.timeout)
+
+      response should be(servingResponse)
+    }
+
+    "report NOT_SERVING when querying a single, unhealthy component" in {
+      val service = new GrpcHealthService(new HealthChecks("component" -> unhealthyComponent))
+
+      val response =
+        Await.result(service.check(serviceRequestFor("component")), patienceConfig.timeout)
+
+      response should be(notServingResponse)
+    }
+
+    "report SERVING when querying a healthy component alongside other, unhealthy components" in {
+      val service = new GrpcHealthService(
+        new HealthChecks(
+          "component A" -> healthyComponent,
+          "component B" -> healthyComponent,
+          "component C" -> unhealthyComponent,
+        ))
+
+      val response =
+        Await.result(service.check(serviceRequestFor("component B")), patienceConfig.timeout)
+
+      response should be(servingResponse)
+    }
+
+    "report NOT_SERVING when querying an unhealthy component alongside other, healthy components" in {
+      val service = new GrpcHealthService(
+        new HealthChecks(
+          "component A" -> unhealthyComponent,
+          "component B" -> healthyComponent,
+          "component C" -> healthyComponent,
+        ))
+
+      val response =
+        Await.result(service.check(serviceRequestFor("component A")), patienceConfig.timeout)
+
+      response should be(notServingResponse)
+    }
+
     "observe changes in health" in {
       val responseObserver = new MockServerCallStreamObserver[HealthCheckResponse]
-      val service = new GrpcHealthService()
 
-      service.watch(request, responseObserver)
-      responseObserver.demandResponse()
+      var componentAHealth: HealthStatus = Healthy
+      var componentBHealth: HealthStatus = Healthy
+      var componentCHealth: HealthStatus = Healthy
+      val service = new GrpcHealthService(
+        new HealthChecks(
+          "component A" -> new StubReporter(componentAHealth),
```

Comment:
Could do. I'll experiment.

Reply:
To be clear, I don't consider it critical, but was thinking a bit about it, because it wasn't immediately obvious what was going on. Maybe changing the by-name parameter to a more explicit `Unit => Health` is already enough without the extra shenanigans.

Reply:
Now using `StubReporter.fixed(Healthy)` or `StubReporter.changing(() => currentHealth)`.